docs/proposals/22-autoscaling-kube-apiserver-via-independent-hpa-and-vpa.md

```diff
@@ -0,0 +1,373 @@
+Then consumption changes to 5 cores, leading HPA to recommend 5 single-core replicas - a change of +4 replicas. VPA
+suggests adding 4 cores to the existing single replica. A HVPA configured at 50% HPA + 50% VPA, would apply half of
+each recommendation, +2 cores and +2 replicas, resulting in (1 + 2) replicas * (1 + 2) cores. Failing to account for
+the multiplicative interaction between changes, HVPA provides a total of 9 cores to a workload which only needs 5. 
+
+### Other Benefits to Replacing the Existing Solution
+- HVPA applies recommendations via direct edits to the pod objects. This approach requires nonobvious special handling
+  by Gardener code outside HVPA, which is error-prone.
+
+### Goals
+- Improve ShootKapi autoscaling stability by eliminating the stability issues existing in the current HVPA-based solution.
+- Improve ShootKapi operational efficiency at low application load levels, by avoiding excessive replicas driven by
+  HVPA's policy for early horizontal scale-out.
+- Allow usage of VPA for right-sizing the resource requests during the entire time and not only under certain conditions.
+- Enable future migration away from HVPA by removing all dependency on it for scaling ShootKapi.
+- Reduce Gardener custom code base and ongoing maintenance effort.
+
+### Non-Goals
+- There are other components currently scaled via HVPA. Changing the autoscaling mode for these is outside the scope
+  of this GEP.
+- Changing the autoscaling of `kube-apiserver`s other than the one in the shoot control plane, is outside the scope of
+  this GEP. These `kube-apiserver`s are subject to a different mode of autoscaling (e.g. vertical-only, where HVPA is
+  only used to stabilize VPA behavior), and it is advantageous from project execution perspective to address these as a 
+  separate concern.
+
+## Proposal
+ShootKapi HVPA is replaced by individual HPA and VPA. Undesirable interference between HPA and VPA is avoided through
+the use of sufficiently independent driving signals. HPA is driven by the rate of HTTP requests to kube-apiserver. VPA is
+driven by CPU and memory usage metrics.
+
+### Rationale
+A rough, heuristic horizontal sizing sets the stage for optimal VPA operation (horizontal size is within stable and
+reasonably efficient range). Vertical scaling fine-tunes requests, ensuring high efficiency (requests not too high)
+and stability (requests not too low).
+
+### Design Outline
+One HPA and one VPA resources are deployed alongside each ShootKapi deployment object. The VPA controls CPU and memory
+requests but not limits. The HPA is driven by the average rate of API requests per pod of the target kube-apiserver
+deployment.
+
+An API request rate metric is deployed to shoot prometheus (one per shoot). It gets aggregated by the aggregate
+prometheus of the respective seed (one per seed) through Gardener's existing metrics aggregation mechanism.
```

Comment:
>This doesn't sound like a sound signal path for autoscaling.

I wouldn't go as far as "unsound", but similar concerns are expressed in the "Heavyweight Prometheus infrastructure feeds signal back into cluster control loop" section. Overall, I did not see the extra latency causing availability problems during tests.

Theoretically, it should be possible, so it's just that tests did not create the specific conditions where latency would surface. But I also expect that the frequency of it happening in real world use will be relatively low. I imagine it would require a sharp spike in request rate, at a point where the request count is just below the horizontal scaling threshold, and the request profile is heavily biased toward costlier requests (as to exhaust the horizontal safety margin).

It's perfectly possible, but then it's also possible to e.g. overwhelm VPA with a sharp load spike. Do you know of a client timeout which would be exceeded specifically due to the extra latency? That the latency would be crossing some threshold where some self-remediation action would be abandoned?

On a related note, deploying a separate prometheus to serve autoscaling seems prohibitively costly at this point. I am investigating into the option of using the cache prometheus for his.